File: packages/admob_kit/ios_integration/NativeAdFactoryExample.swift
// Copy this file into your app's ios/Runner/ directory.
//
// Renders a loaded google_mobile_ads NativeAd as a plain UIKit view. Edit
// the layout inside `createNativeAd` to change the native ad's design -
// this class and the Dart `AdNative` widget never need to change for a
// pure layout tweak. For a more complex design, replace the programmatic
// layout below with a XIB-backed view.

import google_mobile_ads

class NativeAdFactoryExample: FLTNativeAdFactory {

    func createNativeAd(
        _ nativeAd: GADNativeAd,
        customOptions: [AnyHashable: Any]? = nil
    ) -> GADNativeAdView? {
        let adView = GADNativeAdView(frame: .zero)

        let headline = UILabel()
        headline.font = .boldSystemFont(ofSize: 16)
        headline.text = nativeAd.headline
        adView.headlineView = headline

        let body = UILabel()
        body.font = .systemFont(ofSize: 13)
        body.numberOfLines = 2
        body.text = nativeAd.body
        body.isHidden = nativeAd.body == nil
        adView.bodyView = body

        let icon = UIImageView(image: nativeAd.icon?.image)
        icon.isHidden = nativeAd.icon == nil
        adView.iconView = icon

        let cta = UIButton(type: .system)
        cta.setTitle(nativeAd.callToAction, for: .normal)
        cta.isUserInteractionEnabled = false
        cta.isHidden = nativeAd.callToAction == nil
        adView.callToActionView = cta

        let stack = UIStackView(arrangedSubviews: [icon, headline, body, cta])
        stack.axis = .horizontal
        stack.alignment = .center
        stack.spacing = 8
        stack.translatesAutoresizingMaskIntoConstraints = false

        adView.addSubview(stack)
        NSLayoutConstraint.activate([
            stack.leadingAnchor.constraint(equalTo: adView.leadingAnchor, constant: 12),
            stack.trailingAnchor.constraint(equalTo: adView.trailingAnchor, constant: -12),
            stack.topAnchor.constraint(equalTo: adView.topAnchor, constant: 8),
            stack.bottomAnchor.constraint(equalTo: adView.bottomAnchor, constant: -8),
        ])

        adView.nativeAd = nativeAd
        return adView
    }
}

// ---------------------------------------------------------------------
// Registration - add to ios/Runner/AppDelegate.swift:
//
// import google_mobile_ads
//
// @main
// @objc class AppDelegate: FlutterAppDelegate {
//   override func application(
//     _ application: UIApplication,
//     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
//   ) -> Bool {
//     let factory = NativeAdFactoryExample()
//     FLTGoogleMobileAdsPlugin.registerNativeAdFactory(
//       self,
//       factoryId: "adFactoryExample", // must match AdMobConfig.nativeAdFactoryId
//       nativeAdFactory: factory
//     )
//     GeneratedPluginRegistrant.register(with: self)
//     return super.application(application, didFinishLaunchingWithOptions: launchOptions)
//   }
// }
// ---------------------------------------------------------------------

File: packages/admob_kit/android_integration/NativeAdFactoryExample.kt
package com.example.yourapp

// Copy this file into your app's
// android/app/src/main/kotlin/<your/package/path>/ and adjust the
// `package` line above to match. Also copy native_ad_layout.xml into
// android/app/src/main/res/layout/.
//
// This factory turns a loaded google_mobile_ads NativeAd into the Android
// view defined by native_ad_layout.xml. Edit that layout file (and the
// bindings below) to change the native ad's design - this class and the
// Dart `AdNative` widget never need to change for a pure layout tweak.

import android.content.Context
import android.view.LayoutInflater
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.NativeAdFactory

class NativeAdFactoryExample(private val context: Context) : NativeAdFactory {

    override fun createNativeAd(
        nativeAd: NativeAd,
        customOptions: MutableMap<String, Any>?,
    ): NativeAdView {
        val adView = LayoutInflater.from(context)
            .inflate(R.layout.native_ad_layout, null) as NativeAdView

        val headline = adView.findViewById<android.widget.TextView>(R.id.ad_headline)
        val body = adView.findViewById<android.widget.TextView>(R.id.ad_body)
        val icon = adView.findViewById<android.widget.ImageView>(R.id.ad_app_icon)
        val cta = adView.findViewById<android.widget.Button>(R.id.ad_call_to_action)

        headline.text = nativeAd.headline
        adView.headlineView = headline

        if (nativeAd.body != null) {
            body.text = nativeAd.body
            body.visibility = android.view.View.VISIBLE
        } else {
            body.visibility = android.view.View.INVISIBLE
        }
        adView.bodyView = body

        if (nativeAd.icon != null) {
            icon.setImageDrawable(nativeAd.icon!!.drawable)
            icon.visibility = android.view.View.VISIBLE
        } else {
            icon.visibility = android.view.View.GONE
        }
        adView.iconView = icon

        if (nativeAd.callToAction != null) {
            cta.text = nativeAd.callToAction
            cta.visibility = android.view.View.VISIBLE
        } else {
            cta.visibility = android.view.View.INVISIBLE
        }
        adView.callToActionView = cta

        adView.setNativeAd(nativeAd)
        return adView
    }
}

// ---------------------------------------------------------------------
// Registration - add to your app's MainActivity.kt:
//
// import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin
//
// class MainActivity : FlutterActivity() {
//     override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
//         super.configureFlutterEngine(flutterEngine)
//         GoogleMobileAdsPlugin.registerNativeAdFactory(
//             flutterEngine,
//             "adFactoryExample", // must match AdMobConfig.nativeAdFactoryId
//             NativeAdFactoryExample(context)
//         )
//     }
//
//     override fun cleanUpFlutterEngine(flutterEngine: FlutterEngine) {
//         super.cleanUpFlutterEngine(flutterEngine)
//         GoogleMobileAdsPlugin.unregisterNativeAdFactory(
//             flutterEngine,
//             "adFactoryExample"
//         )
//     }
// }
// ---------------------------------------------------------------------
